### ISSUE TITLE
Lint/Void maybe not working right in setter methods

### ISSUE THREAD

**Issue by will** [rubocop-hq/rubocop#8811]
Given the following file

``` ruby
# frozen_string_literal: true

# A comment
class Foo
  def some_setter=(val)
    @val = run_some_code_and_change_type(val)
    val
  end

  def run_some_code_and_change_type(val)
    val.to_s
  end
end
```

## Expected behavior

There should be no Lint/Void error on `val` on the last line of the method `#some_setter=` since I'm using it as a return value (and also I've used in in a method call it to set `@val`)

## Actual behavior

```
Inspecting 1 file
W

Offenses:

test.rb:7:5: W: Lint/Void: Variable val used in void context.
    val
    ^^^

1 file inspected, 1 offense detected, 1 offense auto-correctable
```

## RuboCop version

```
$ rubocop -V
0.92.0 (using Parser 2.7.1.5, rubocop-ast 0.7.0, running on ruby 2.6.6 x86_64-darwin19)
```

Not sure if it's strictly relevant, but I discovered this through standardrb

```
$ standardrb -v
0.7
```

and reported it there first: https://github.com/testdouble/standard/issues/206

**Comment by will** [comment]
Also notice that the output says it's auto-correctable. Trying to actually autocorrect it fails. Also every once in a while it wont say `1 offense auto-correctable`, but it usually does, even if the file hasn't changed at all.

**Comment by tejasbubane** [comment]
I think the offense is correct. Assignment methods should return the assigned value (for which they don't need explicit return as assignment will return it). 

Assignment methods are explicitly marked as `void context` [here](https://github.com/rubocop-hq/rubocop-ast/blob/210ce65e91e50466fba4efe46891ad82f00bf242/lib/rubocop/ast/node/def_node.rb#L16).

**Comment by will** [comment]
Do they return the value that is assigned, or the value that is passed in? I can't think of places to check for precedent of methods that modify on the way in and return that modified value.

**Comment by marcandre** [comment]
It normally does not matter what you return

```ruby
obj = Object.new; def obj.foo=(_); 42;end
x = obj.foo = 666
x # => 666, not 42
```

The only way to get the result is to use `send :foo=` and use the result; nobody should do that.

**Comment by marcandre** [comment]
The "1 offense auto-correctable" seems to be a bug though. `Lint/Void` doesn't even `extend AutoCorrector`, odd.

**Comment by marcandre** [comment]
Can't reproduce the "1 offense auto-correctable" with 0.92.0 or master, so unless there's anything else, I'll close this, but we can reopen if there's something else.

Just to be clear, the correct code for your example should simply be:

```ruby
# A comment
class Foo
  def some_setter=(val)
    @val = run_some_code_and_change_type(val)
  end

  def run_some_code_and_change_type(val)
    val.to_s
  end
end
